File: tutorials/deeplearning/deeplearning.R
library(h2o)
args(h2o.deeplearning)
help(h2o.deeplearning)
example(h2o.deeplearning)
#demo(h2o.deeplearning)
h2o.init(nthreads=-1, max_mem_size = "2G")
spiral.hex <- h2o.importFile("/users/arno/h2o-world-2015-training/tutorials/data/spiral.csv")
spiral <- as.data.frame(spiral.hex)
grid.hex <- h2o.importFile("/users/arno/h2o-world-2015-training/tutorials/data/grid.csv")

# Helper to plot the training data and the probability contours obtained from the grid frame
plotContour <- function(name, model, data, grid.hex) {
  n=0.5*(sqrt(nrow(grid.hex))-1); d <- 1.5; h <- d*(-n:n)/n
  pred <- as.data.frame(h2o.predict(model, grid.hex))  ##make predictions on a square grid
  plot(data[,-3],pch=19,col=data[,3],cex=1.0,xlim=c(-d,d),ylim=c(-d,d),main=name) ##plot data
  contour(h,h,z=array(ifelse(pred[,1]=="Red",0,1),dim=c(2*n+1,2*n+1)),col="blue",lwd=2,add=T)
}
# set up the canvas for 2x2 plots
par(mfrow=c(2,2))
plotContour("DL",  h2o.deeplearning(x=1:2,y=3,training_frame=spiral.hex,epochs=1000), spiral, grid.hex)
plotContour("GBM", h2o.gbm(x=1:2,y=3,training_frame=spiral.hex), spiral, grid.hex)
plotContour("DRF", h2o.randomForest(x=1:2,y=3,training_frame=spiral.hex), spiral, grid.hex)
plotContour("GLM", h2o.glm(x=1:2,y=3,training_frame=spiral.hex, family="binomial"), spiral, grid.hex)
df <- h2o.importFile("/users/arno/h2o-world-2015-training/tutorials/data/covtype.full.csv")
dim(df)
df
splits <- h2o.splitFrame(df, c(0.6,0.2), seed=1234)
train <- h2o.assign(splits[[1]], "train.hex")
valid <- h2o.assign(splits[[2]], "valid.hex")
test <- h2o.assign(splits[[3]], "test.hex")
plot(h2o.tabulate(df, "Elevation",                       "Cover_Type"))
plot(h2o.tabulate(df, "Horizontal_Distance_To_Roadways", "Cover_Type"))
plot(h2o.tabulate(df, "Soil_Type",                       "Cover_Type"))
plot(h2o.tabulate(df, "Horizontal_Distance_To_Roadways", "Elevation"))
plot(h2o.tabulate(df, "Horizontal_Distance_To_Roadways", "Elevation"))
response <- "Cover_Type"
predictors <- setdiff(names(df), response)
predictors
m <- h2o.deeplearning(
  model_id="dl_model_first", 
  training_frame=train, 
  validation_frame=valid,   ## validation dataset: used for scoring and early stopping
  x=predictors,
  y=response,
  #activation="Rectifier",  ## default
  #hidden=c(200,200),       ## default: 2 hidden layers with 200 neurons each
  epochs=1,
  variable_importances=T    ## not enabled by default
)
summary(m)
vi <- as.data.frame(h2o.varimp(m))
hist(vi$relative_importance)
m <- h2o.deeplearning(
  model_id="dl_model_faster", 
  training_frame=train, 
  validation_frame=valid,
  x=predictors,
  y=response,
  hidden=c(32,32,32),               ## small network, runs faster
  epochs=1000000,
  score_validation_samples = 10000, ## sample the valiation dataset, faster and accurate enough
  stopping_rounds=1,
  stopping_metric="misclassification",  ## could be "MSE","logloss","r2"
  stopping_tolerance=0.01
)
summary(m)
plot(m)

## with some tuning: ~8% in 40s
m <- h2o.deeplearning(
  model_id="dl_model_tuned", 
  training_frame = train, 
  validation_frame = valid, 
  x=predictors, 
  y=response, 
  overwrite_with_best_model=F,
  hidden=c(100,100,100),          ## more hidden layers -> more complex interactions
  epochs=10,                     ## long enough to converge
  score_validation_samples=10000, ## downsample validation set for faster scoring
  score_duty_cycle=0.025,         ## don't score more than 2.5% of the wall time
  adaptive_rate=F,                ## manually tuned learning rate
  rate=0.02, 
  rate_annealing=2e-6,            
  momentum_start = 0.2,           ## manually tuned momentum
  momentum_stable = 0.4, 
  momentum_ramp = 1e7, 
  l1=1e-5,                        ## add some L1/L2 regularization
  l2=1e-5,
  max_w2 = 10                     ## helps stability for Rectifier
) 
summary(m)
h2o.confusionMatrix(h2o.performance(m, train=T)) ## training data
h2o.confusionMatrix(h2o.performance(m, valid=T)) ## sampled validation data
h2o.confusionMatrix(m, valid)                    ## validation datas
h2o.confusionMatrix(m, test)                     ## test data
p <- h2o.predict(m, test)
p
test$Accuracy <- p$predict == test$Cover_Type
1-mean(test$Accuracy)
dlmodel@model$params 
  sampled_train = train[1:10000,]
hyper_params <- list(
  hidden = list(c(32,32,32),c(64,64)),
  input_dropout_ratio = c(0,0.05),
  hidden_dropout_ratios = list(c(0,0,0),c(0.05,0.05)), ## only one will match a given 'hidden' parameter
  rate=c(0.01,0.02),
  rate_annealing=c(1e-8,1e-7,1e-6)
)
hyper_params
grid <- h2o.grid(
  "deeplearning",
  model_id="dl_grid", 
  training_frame = sampled_train,
  validation_frame = valid, 
  x=predictors, 
  y=response,
  epochs=10,
  stopping_metric="misclassification",
  stopping_tolerance=1e-2,        ## stop when logloss does not improve by >=1% for 2 scoring events
  stopping_rounds=2,
  score_validation_samples=10000, ## downsample validation set for faster scoring
  score_duty_cycle=0.025,         ## don't score more than 2.5% of the wall time
  adaptive_rate=F,                ## manually tuned learning rate
  momentum_start = 0.5,           ## manually tuned momentum
  momentum_stable = 0.9, 
  momentum_ramp = 1e7, 
  l1 = 1e-5,
  l2 = 1e-5,
  activation = c("RectifierWithDropout"),
  max_w2 = 10,                    ## can help improve stability for Rectifier
  hyper_params = hyper_params
)
## Find the best model and its full set of parameters (clunky for now, will be improved)
scores <- cbind(as.data.frame(unlist((lapply(grid@model_ids, function(x) 
  { h2o.confusionMatrix(h2o.performance(h2o.getModel(x),valid=T))$Error[8] })) )), unlist(grid@model_ids))
best_err <- scores[1,1]
best_model <- h2o.getModel(as.character(scores[1,2]))
print(paste0("Best misclassification: ", best_err))
print(best_model@allparameters)
models <- c()
for (i in 1:10) {
  rand_activation <- c("TanhWithDropout", "RectifierWithDropout")[sample(1:2,1)]
  rand_numlayers <- sample(2:5,1)
  rand_hidden <- c(sample(10:50,rand_numlayers,T))
  rand_l1 <- runif(1, 0, 1e-3)
  rand_l2 <- runif(1, 0, 1e-3)
  rand_dropout <- c(runif(rand_numlayers, 0, 0.6))
  rand_input_dropout <- runif(1, 0, 0.5)
  dlmodel <- h2o.deeplearning(
    model_id=paste0("dl_random_model_", i),
    training_frame = sampled_train,
    validation_frame = valid, 
    x=predictors, 
    y=response,
    epochs=10,
    stopping_metric="misclassification",
    stopping_tolerance=1e-2,        ## stop when logloss does not improve by >=1% for 2 scoring events
    stopping_rounds=2,
    score_validation_samples=10000, ## downsample validation set for faster scoring
    score_duty_cycle=0.025,         ## don't score more than 2.5% of the wall time
    max_w2 = 10,                    ## can help improve stability for Rectifier

    ### Random parameters
    activation=rand_activation, 
    hidden=rand_hidden, 
    l1=rand_l1, 
    l2=rand_l2,
    input_dropout_ratio=rand_input_dropout, 
    hidden_dropout_ratios=rand_dropout
  )                                
  models <- c(models, dlmodel)
}
if (is.null(best_err)) best_err <- 1      ##use the best reference model from the grid search above
for (i in 1:length(models)) {
  err <- h2o.confusionMatrix(h2o.performance(models[[i]],valid=T))$Error[8]
  if (err < best_err) {
    best_err <- err
    best_model <- models[[i]]
  }
}
best_model
best_params <- best_model@parameters
best_params$activation
best_params$hidden
best_params$l1
best_params$l2
best_params$input_dropout_ratio
best_params$hidden_dropout_ratios
#max_epochs <- 1000 ##Takes a few minutes
max_epochs <- 20    ##Takes about 30s
m <- h2o.deeplearning(
  model_id="dl_model_tuned_continued", 
  checkpoint="dl_model_tuned", 
  training_frame = train, 
  validation_frame = valid, 
  x=predictors, 
  y=response, 
  hidden=c(100,100,100),          ## more hidden layers -> more complex interactions
  epochs=max_epochs,              ## hopefully long enough to converge (otherwise restart again)
  stopping_metric="logloss",
  stopping_tolerance=1e-2,        ## stop when validation logloss does not improve by >=1% for 2 scoring events
  stopping_rounds=2,
  score_validation_samples=10000, ## downsample validation set for faster scoring
  score_duty_cycle=0.025,         ## don't score more than 2.5% of the wall time
  adaptive_rate=F,                ## manually tuned learning rate
  rate=0.02, 
  rate_annealing=2e-6,            
  momentum_start = 0.2,           ## manually tuned momentum
  momentum_stable = 0.4, 
  momentum_ramp = 1e7, 
  l1=1e-5,                        ## add some L1/L2 regularization
  l2=1e-5,
  max_w2 = 10                     ## helps stability for Rectifier
) 
summary(m)
plot(m)
h2o.saveModel(dlmodel_continued, dir="/tmp", name="mybest_covtype_model", force=T)
dlmodel_loaded <- h2o.loadModel(h2oServer, "/tmp/mybest_covtype_model")
dlmodel_continued_again <- h2o.deeplearning(x=c(1:784), y=785, data=train_hex, validation=test_hex,
                            checkpoint = dlmodel_loaded, l1=best_params$l1, epochs=0.5)

dlmodel_continued_again@model$valid_class_error
dlmodel <- h2o.deeplearning(
  x=predictors,
  y=response, 
  training_frame=train,
  hidden=c(10,10),
  epochs=0.1,
  nfolds=5)
summary(dlmodel)
    dlmodel <- h2o.deeplearning(x=1:784, y=785, data=train_hex, validation=test_hex,
                                hidden=c(50,50), epochs=0.1, activation="Tanh",
                                score_training_samples=1000, score_validation_samples=1000,
                                balance_classes=TRUE,
                                score_validation_sampling="Stratified")

File: tutorials/deeplearning/deeplearning.py

# coding: utf-8

# # Introduction
# This tutorial shows how a H2O [Deep Learning](http://en.wikipedia.org/wiki/Deep_learning) model can be used to do supervised classification and regression. This tutorial covers usage of H2O from R. A python version of this tutorial will be available as well in a separate document. This file is available in plain R, R markdown and regular markdown formats, and the plots are available as PDF files. More examples and explanations can be found in our [H2O Deep Learning booklet](http://h2o.ai/resources/) and on our [H2O Github Repository](http://github.com/h2oai/h2o-3/).
# 

# ### H2O Python Module
# 
# Load the H2O Python module.

# In[1]:

import h2o


# ### Start H2O
# Start up a 1-node H2O server on your local machine, and allow it to use all CPU cores and up to 2GB of memory:

# In[2]:

h2o.init(max_mem_size_GB = 2)            #uses all cores by default
h2o.remove_all()                          #clean slate, in case cluster was already running


# To learn more about the h2o package itself, we can use Python's builtin help() function.

# In[3]:

help(h2o)


# help() can be used on H2O functions and models. Jupyter's builtin shift-tab functionality also works

# In[4]:

from h2o.estimators.deeplearning import H2OAutoEncoderEstimator, H2ODeepLearningEstimator
help(H2ODeepLearningEstimator)
help(h2o.import_file)


# ##H2O Deep Learning
# While H2O Deep Learning has many parameters, it was designed to be just as easy to use as the other supervised training methods in H2O. Early stopping, automatic data standardization and handling of categorical variables and missing values and adaptive learning rates (per weight) reduce the amount of parameters the user has to specify. Often, it's just the number and sizes of hidden layers, the number of epochs and the activation function and maybe some regularization techniques.
# 

# ### Let's have some fun first: Decision Boundaries
# We start with a small dataset representing red and black dots on a plane, arranged in the shape of two nested spirals. Then we task H2O's machine learning methods to separate the red and black dots, i.e., recognize each spiral as such by assigning each point in the plane to one of the two spirals.

# We visualize the nature of H2O Deep Learning (DL), H2O's tree methods (GBM/DRF) and H2O's generalized linear modeling (GLM) by plotting the decision boundary between the red and black spirals:

# In[5]:

#get_ipython().magic(u'matplotlib inline')
#IMPORT ALL THE THINGS

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from h2o.estimators.deeplearning import H2OAutoEncoderEstimator, H2ODeepLearningEstimator
from h2o.estimators.gbm import H2OGradientBoostingEstimator
from h2o.estimators.glm import H2OGeneralizedLinearEstimator
from h2o.estimators.random_forest import H2ORandomForestEstimator


# First, we need to upload our datasets to the the H2O cluster. The data is imported into H2OFrames, which operate similarly in function to pandas DataFrames.  
# 
# In this case, the cluster is running on our laptops. Data files are imported by their relative locations to this notebook.

# In[6]:

spiral = h2o.upload_file("../data/spiral.csv")
grid  = h2o.upload_file("../data/grid.csv")


# Spiral is a simple data set consisting of two spirals of black and red dots.  
# Grid is a 201 by 201 matrix with dimensions [-1.5, 1.5] by [-1.5, 1.5].
# 
# To visualize these datasets, we can pull them from H2OFrames into pandas DataFrames for easier plotting.

# In[7]:

spiral_df = spiral.as_data_frame(use_pandas=True)
grid_df = grid.as_data_frame(use_pandas=True)
grid_x, grid_y = grid_df.x.reshape(201,201), grid_df.y.reshape(201,201)
spiral_r = spiral_df[spiral_df.color == "Red"]
spiral_k = spiral_df[spiral_df.color == "Black"]

spiral_xr, spiral_yr = spiral_r[spiral_r.columns[0]], spiral_r[spiral_r.columns[1]]
spiral_xk, spiral_yk = spiral_k[spiral_k.columns[0]], spiral_k[spiral_k.columns[1]]
    
markersize_ = 7**2
plt.figure(figsize = (5,5))
plt.scatter(spiral_xr, spiral_yr, c = 'r', s=markersize_)
plt.scatter(spiral_xk, spiral_yk, c = 'k', s=markersize_)
plt.axis([-1.5, 1.5, -1.5, 1.5])
plt.title("Spiral");


# ###Model Construction
# H2O in Python is designed to be very similar in look and feel to to scikit-learn. Models are initialized individually with desired or default parameters and then trained on data.  
# 
# Note that the below examples use model.train(), as opposed the traditional model.fit()
# This is because h2o-py takes the data frame AND column indices for the feature and response columns, while scikit-learn takes in feature frames.
# 
# H2O supports model.fit() so that it can be incorporated into a scikit-learn pipeline, but we advise using train() in all other cases.

# In[8]:

X = spiral.col_names[0:2]
y = spiral.col_names[2]
dl_model = H2ODeepLearningEstimator(epochs=1000)
dl_model.train(X, y, spiral)

gbm_model = H2OGradientBoostingEstimator()
gbm_model.train(X, y, spiral)

drf_model = H2ORandomForestEstimator()
drf_model.train(X, y, spiral)

glm_model = H2OGeneralizedLinearEstimator(family="binomial")
glm_model.fit(spiral[X], spiral[y])                                #model.fit() example

models = [dl_model, gbm_model, drf_model, glm_model]
m_names = ["Deep Learning", "Gradient Boosted Method", "Distributed Random Forest", "Generalized Linear Model"]


# Now that we've trained four models to classify points as red or black based on their (x,y) coordinates.  
# To see how our models have performed, we ask them to predict the colors of the grid.
# 
# Since we'll be doing a lot of spiral plotting, let's write a little helper function to keep things clean.

# In[79]:

def plot_spirals(models, model_names):
    fig, ax = plt.subplots(2,2, figsize=(12,12))
    for k, subplot in enumerate(ax.flatten()):
        subplot.scatter(spiral_xr, spiral_yr, c = 'r', s=markersize_)
        subplot.scatter(spiral_xk, spiral_yk, c = 'k', s=markersize_)
        subplot.axis([-1.5, 1.5, -1.5, 1.5])
        subplot.set_title(model_names[k])
        subplot.set_xlabel('x')
        subplot.set_ylabel('y')
        pred_z = models[k].predict(grid).as_data_frame(True)
        subplot.contour(grid_x, grid_y, (pred_z['predict'] == 'Black').astype(np.int).reshape(201,201), colors='b')


# Below are four graphs of the contour plots of the predictions, so that we can see how exactly the algorithms grouped the points into black and red.

# In[37]:

plot_spirals(models, m_names)


# ###A Deeper Dive into Deep Learning
# 
# Now let's explore the evolution of our deep learning model over training time (number of passes over the data, aka epochs).  
# We will use checkpointing to ensure that we continue training the same model

# In[77]:

dl_1 = H2ODeepLearningEstimator(epochs=1)
dl_1.train(X, y, spiral)

dl_250 = H2ODeepLearningEstimator(checkpoint=dl_1, epochs=250)
dl_250.train(X, y, spiral)

dl_500 = H2ODeepLearningEstimator(checkpoint=dl_250, epochs=500)
dl_500.train(X, y, spiral)

dl_750 = H2ODeepLearningEstimator(checkpoint=dl_500, epochs=750)
dl_750.train(X, y, spiral)


# You can see how the network learns the structure of the spirals with enough training time. 

# In[80]:

models_dl = [dl_1, dl_250, dl_500, dl_750]
m_names_dl = ["DL " + str(int(model.get_params()['epochs']['actual_value'])) +                                      " Epochs" for model in models_dl]

plot_spirals(models_dl, m_names_dl)


# ###Deep Learning Network Architecture
# Of course, there is far more to constructing Deep Learning models than simply having them run longer.  
# Consider the four following setups.
# 
# 1. Single layer, 1000 nodes
# 2. Two layers, 200 nodes each
# 3. Three layers, 42 nodes each
# 4. Four layers, 11 -> 13 -> 17 -> 19
# 
# The H2O Architecture uses the hidden keyword to control model network architecture.  
# Hidden takes a list of integers, representing the number of nodes in each layer.

# In[30]:

dl_1 = H2ODeepLearningEstimator(hidden=[1000], epochs=500)
dl_1.train(X, y, spiral)

dl_2 = H2ODeepLearningEstimator(hidden=[200,200], epochs=500)
dl_2.train(X, y, spiral)

dl_3 = H2ODeepLearningEstimator(hidden=[42,42,42], epochs=500)
dl_3.train(X, y, spiral)

dl_4 = H2ODeepLearningEstimator(hidden=[11,13,17,19], epochs = 1000)
dl_4.train(X, y, spiral)


# It is clear that different configurations can achieve similar performance, and that tuning will be required for optimal performance.

# In[31]:

models_network = [dl_1, dl_2, dl_3, dl_4]
m_names_network = ["1000", "200 x 200", "42 x 42 x 42", "11 x 13 x 17 x 19"]

plot_spirals(models_network, m_names_network)


# ###Activation Functions
# Next, we compare between different activation functions, including one with 50% dropout regularization in the hidden layers:

# In[33]:

models_act = []
m_names_act = []
for i,method in enumerate(["Tanh","Maxout","Rectifier","RectifierWithDropout"]):
    models_act.append(H2ODeepLearningEstimator(activation=method, hidden=[100,100], epochs=1000))
    models_act[i].train(X, y, spiral)
    m_names_act.append("DL "+ method + " Activation")


# In[35]:

plot_spirals(models_act, m_names_act)


# Clearly, the dropout rate was too high or the number of epochs was too low for the last configuration, which often ends up performing the best on larger datasets where generalization is important.  
# 
# More information about the parameters can be found in the [H2O Deep Learning booklet](http://h2o.ai/resources/).

# ## Covertype Dataset
# The following examples use the Covertype dataset from UC Irvine, which concerns predicting forest cover based on cartographical data.  
# We import the full covertype dataset (581k rows, 13 columns, 10 numerical, 3 categorical) and then split the data 3 ways:  
#   
# 60% for training  
# 20% for validation (hyper parameter tuning)  
# 20% for final testing  

# In[85]:

covtype_df = h2o.import_file("../data/covtype.full.csv")

#split the data as described above
train, valid, test = covtype_df.split_frame([0.6, 0.2], seed=1234)

#Prepare predictors and response columns
covtype_X = covtype_df.col_names[:-1]     #last column is cover_type, 
covtype_y = covtype_df.col_names[-1]    


# ####First Impressions
# Let's run our first Deep Learning model on the covtype dataset.   
# We want to predict the `Cover_Type` column, a categorical feature with 7 levels, and the Deep Learning model will be tasked to perform (multi-class) classification. It uses the other 12 predictors of the dataset, of which 10 are numerical, and 2 are categorical with a total of 44 levels.  
# 
# We can expect the Deep Learning model to have 56 input neurons (after automatic one-hot encoding). First run will be only one epoch to get a feel for the model construction.

# In[88]:

#set the model_id for easy lookup in Flow
covtype_model_v1 = H2ODeepLearningEstimator(model_id="covtype_v1", epochs=1, variable_importances=True)
covtype_model_v1.train(covtype_X, covtype_y, training_frame = train, validation_frame = valid)
print covtype_model_v1


# ###Shutdown H2O Cluster
# Shut down the cluster now that we are done using it.

# In[ ]:

h2o.shutdown()

